### PR TITLE
Capture VM's console if VM hangs on boot or poweroff

### PR DESCRIPTION
**Pull request by skobyda**
* [ ] FAIL: image-refresh centos-9-stream

### CODE REVIEW

**Comment by skobyda** [comment]
Tried to test it locally with currently failing "rhel-9-3", it managed to capture the console log successfully:
https://gist.githubusercontent.com/skobyda/1b85557c87ad1d91be954e1e00b9f42a/raw/47c9359805184bf3c2a5283184d413213e79eda3/gistfile1.txt

File: machine/machine_core/ssh_connection.py [hunk 1]
```diff
@@ -69,6 +69,22 @@ def message(self, *args):
         if self.verbose:
             sys.stderr.write(" ".join(args) + '\n')
 
+    def print_console_log(self):
+        """Prints VM's console to stderr"""
+        console_file = "/tmp/console.log"
+        try:
+            f = open(console_file)
+            console_log = f.read().strip()
```
Comment (martinpitt) [inline machine/machine_core/ssh_connection.py]: Please reduce the scope of the try/except to the open/read pair, and catch all `OSError`s -- this is already in an exception handler, so this better not crash.  Print the exception when it happens for debugging.

Please also use a `with` context (see codeql)

File: machine/machine_core/ssh_connection.py [hunk 2]
```diff
@@ -69,6 +69,22 @@ def message(self, *args):
...
+
+            if len(console_log) == 0:
+                sys.stderr.write(f"VM's console log file '{console_file}' is empty")
+                return
+
+            sys.stderr.write(f"---- Console log start's here ----\n{console_log}\n---- Console log end's here ----")
```
Comment (martinpitt) [inline machine/machine_core/ssh_connection.py]: "starts" and "ends" (no apostrophes)

Also, write() does not automatically append a `\n`. Either do that, or use `print(..., file=sys.stderr)`

File: machine/machine_core/ssh_connection.py [hunk 3]
```diff
@@ -69,6 +69,22 @@ def message(self, *args):
         if self.verbose:
             sys.stderr.write(" ".join(args) + '\n')
 
+    def print_console_log(self):
+        """Prints VM's console to stderr"""
+        console_file = "/tmp/console.log"
```
Comment (martinpitt) [inline machine/machine_core/ssh_connection.py]: This looks wrong. Is this a test leftover? If this path is set anywhere, it won't work on our bots, as they run many VMs in parallel. Also, fixed paths in /tmp are a no-go.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks for figuring this out!

File: machine/machine_core/ssh_connection.py (hunk 3 above)
Comment (skobyda) [reply machine/machine_core/ssh_connection.py]: Fixed

File: machine/machine_core/ssh_connection.py (hunk 1 above)
Comment (skobyda) [reply machine/machine_core/ssh_connection.py]: Fixed

File: machine/machine_core/ssh_connection.py (hunk 2 above)
Comment (skobyda) [reply machine/machine_core/ssh_connection.py]: Fixed

File: machine/machine_core/machine.py [hunk 4]
```diff
@@ -53,8 +53,8 @@
 
 
 class Machine(ssh_connection.SSHConnection):
-    def __init__(self, address="127.0.0.1", image="unknown", verbose=False, label=None, browser=None,
-                 user="root", identity_file=None, arch="x86_64", ssh_port=22, web_port=9090):
+    def __init__(self, address="127.0.0.1", image="unknown", verbose=False, console_file=None, label=None,
+                 browser=None, user="root", identity_file=None, arch="x86_64", ssh_port=22, web_port=9090):
```
Comment (martinpitt) [inline machine/machine_core/machine.py:58]: This is the wrong place. The *only* class where we can do this is VirtMachine. We can't control already running/external/Testing Farm etc. machine types. Please move all of this into VirtMachine and handle the argument there. Machine and SshConnection should not be touched by this.

File: machine/machine_core/ssh_connection.py [hunk 5]
```diff
@@ -69,6 +70,24 @@ def message(self, *args):
         if self.verbose:
             sys.stderr.write(" ".join(args) + '\n')
 
+    def print_console_log(self):
+        """Prints VM's console to stderr"""
+        if not self.console_file:
+            return
+
+        try:
+            f = open(self.console_file)
```
Comment (martinpitt) [inline machine/machine_core/ssh_connection.py:79]: This needs to go inside the `with` already. I. e. `with open(...) as f: log = read()`. The processing of `console_log` shouldn't be in the context.

File: machine/machine_core/ssh_connection.py [hunk 6]
```diff
@@ -69,6 +70,24 @@ def message(self, *args):
...
+            console_log = f.read().strip()
+        except OSError as ex:
+            sys.stderr.write(f"Failed to open '{self.console_file}': {0}".format(ex.strerror))
```
Comment (martinpitt) [inline machine/machine_core/ssh_connection.py:82]: Missing `\n`. Also, please put the exception into the f-string. Combining f-string and .format() is weird.

File: machine/machine_core/ssh_connection.py [hunk 7]
```diff
@@ -69,6 +70,24 @@ def message(self, *args):
...
+
+        with f:
+            if len(console_log) == 0:
+                sys.stderr.write(f"VM's console log file '{self.console_file}' is empty")
```
Comment (martinpitt) [inline machine/machine_core/ssh_connection.py:86]: Missing `\n`

File: machine/machine_core/machine_virtual.py [hunk 8]
```diff
@@ -330,6 +331,8 @@ def _start_qemu(self):
             "memory_in_mib": self.memory_mb,
             "drive": image_to_use,
             "iso": os.path.join(BOTS_DIR, "machine", "cloud-init.iso"),
+            "console_type": "file" if self.console_file else "pty",
+            "console_source": f"<source path='{self.console_file}'/>" if self.console_file else "",
```
Comment (martinpitt) [inline machine/machine_core/machine_virtual.py:335]: This shouldn't be a separate commit, it belongs into the previous one.

File: image-create [hunk 9]
```diff
@@ -207,7 +207,14 @@ try:
         memory_mb = 3072
     else:
         memory_mb = 2048
-    machine = testvm.VirtMachine(verbose=args.verbose, image=args.image, memory_mb=memory_mb, maintain=True)
+
+    console_file = tempfile.NamedTemporaryFile(suffix='.log', prefix='console-').name
```
Comment (martinpitt) [inline image-create]: Not keeping the `NamedTemporaryFile` object around will immediately close it, and thus also remove the file again. So the only effect of this is that you keep a random name. It's not super critical here, but please don't learn this pattern -- it's actively dangerous in "production" applications. Drop the `.name` here, and pass `console_file.name` to VirtMachine instead, so that the TemporaryFile object survives throughout image-create.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks!

File: image-create [hunk 10]
```diff
@@ -207,7 +207,14 @@ try:
         memory_mb = 3072
     else:
         memory_mb = 2048
-    machine = testvm.VirtMachine(verbose=args.verbose, image=args.image, memory_mb=memory_mb, maintain=True)
+
+    console_file = tempfile.NamedTemporaryFile(suffix='.log', prefix='console-')
```
Comment (martinpitt) [inline image-create]: This gets changed again in the third commit. Please move all the changes there.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! one more cleanup round, then it should be good!

File: image-create (hunk 10 above)
Comment (skobyda) [reply image-create]: But console_file is necessary in second commit. Without it it doesn't work. Or do you mean to squash second and third commit?

Comment (martinpitt) [reply image-create]: If they are so interdependent, then just squash them, yes.

File: image-create [hunk 11]
```diff
@@ -207,7 +208,16 @@ try:
...
     else:
         memory_mb = 2048
-    machine = testvm.VirtMachine(verbose=args.verbose, image=args.image, memory_mb=memory_mb, maintain=True)
+
+    console_file = None
+    if args.capture_console:
+        console_file = tempfile.NamedTemporaryFile(suffix='.log', prefix='console-')
+
+    machine = testvm.VirtMachine(verbose=args.verbose,
+                                 console_file=console_file.name,
```
Comment (martinpitt) [inline image-create:217]: Sorry that I didn't see this earlier, as it was too distributed : This is a somewhat weird API. You stuff in a path, but the caller never actually does anything with it. Instead it's handled all internally. I think it may be cleaner to make this a `capture_console=True` flag, and generate the file name inside of VirtMachine. That also fixes the temp file scope issue that you stumbled over before.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! This is a lot cleaner. I didn't see the API issue before, so one more round?

**Comment by cockpituous** [comment]
Failed. Log: https://cockpit-logs.us-east-1.linodeobjects.com/image-refresh-logs/centos-9-stream-20230907-152222.log

**Review by martinpitt — APPROVED** [review]
Thanks! Image refresh runs now. Assuming that fails with a proper log.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Oh, good that we tested that, see https://cockpit-logs.us-east-1.linodeobjects.com/image-refresh-logs/centos-9-stream-20230907-152222.log

**Comment by cockpituous** [comment]
Failed. Log: https://cockpit-logs.us-east-1.linodeobjects.com/image-refresh-logs/centos-9-stream-20230907-154848.log

**Comment by cockpituous** [comment]
Failed. Log: https://cockpit-logs.us-east-1.linodeobjects.com/image-refresh-logs/centos-9-stream-20230907-161515.log

**Review by martinpitt — APPROVED** [review]
Nice, now it fails in the expected manner!
